### ISSUE TITLE
withI18nRouter HOC for locale subpath routing

### ISSUE THREAD

**Issue by isaachinman** [isaachinman/next-i18next#4]
In the [current react-i18next/nextjs example](https://github.com/i18next/react-i18next/tree/master/example/nextjs), we've created a [wrapped Link component](https://github.com/i18next/react-i18next/blob/master/example/nextjs/components/Link.js) to automatically handle locale subpaths via use of `as`.

This option _works_ but I'd like to explore the possibility of exporting a `withI18nRouter` HOC instead, so that users can wrap their components just like they would with the [NextJs withRouter HOC](https://github.com/zeit/next.js#using-a-higher-order-component).

This suggestion comes from @timneutkens in a comment [here](https://github.com/isaachinman/next-i18next/issues/1#issuecomment-438187263).

Imperative routing and/or routing outside of a React component would be left fully up to the user. We might want to consider exporting the `lngPathDetector` and `lngPathCorrector` utils to aid with that.

**Comment by isaachinman** [comment]
As no one has ever requested this feature, I'm going to close the issue. We can reopen in the future if people start to want a `withI18nRouter` HOC.

**Comment by isaachinman** [comment]
Reopening due to discussion in #143. Seems users want a HOC and/or Router instance after all. Let's discuss.

**Comment by isaachinman** [comment]
@capellini From #143:

> Probably all that will happen prior to a 1.0 release is that we expose lang path utils so that "advanced" users can import them and use them for their own imperative routing.

I think this would just mean exposing `lng-path-corrector`, right?

**Comment by capellini** [comment]
If we also expose our `withConfig` HoC, then they could wrap their components inside of it and get the `config` and `i18n` arguments for `lng-path-corrector`.  Then again, this is also available in their `i18n.js` file, so it's not strictly necessary.

Other than that consideration, then yes -- I believe all we would need prior to a 1.0 release is to expose `lng-path-corrector`.

`lng-path-detector` should not be necessary to expose because users can always put their `server.use/get/etc` calls in their `server.js` file after they call our middleware and all of the `next-i18next` considerations are already handled for them.

**Comment by isaachinman** [comment]
@capellini I think in the end I'd rather expose a wrapped `next/router` although probably in a much simpler way than the approach taken in #149.

After some consideration, I think that `lng-path-corrector` will confuse users, and isn't strictly necessary - we should just handle that internally.

**Comment by capellini** [comment]
@isaachinman Yes, that would be the preferable approach.  I had thought that exposing `lng-path-corrector` would be a good short-term fix to get something out for a 1.0 release.  But in retrospect, it's probably better to expose our own router, even if it takes a bit longer to get to 1.0.

**Comment by minocys** [comment]
@isaachinman @capellini Just a quick status update on #149 

I took a minute to update the tests to more thoroughly check that we are not affecting the underlying NextRouter. I think I've addressed both of your immediate concerns mentioned in the PR.

> although probably in a much simpler way than the approach taken in #149.

Beyond replacing the logic taken from the Link component, I'm not sure what you mean by simpler. Am I missing something? I am happy to make changes to match your specs.
